### DISCUSSION TITLE
Iceberg Tables

### DISCUSSION THREAD

**Discussion by josephniblo** [littleK0i/SnowDDL#81] [Ideas]
Hi, [Iceberg Tables](https://docs.snowflake.com/en/user-guide/tables-iceberg) are now in public preview and would make a really valuable addition to SnowDDL. Is there any plan to add support?

**Comment by littleK0i** [comment]
(edited 18 Dec 2024)

SnowDDL has a limited support for unmanaged Iceberg tables starting with `0.38.0`.

Link to guide: https://docs.snowddl.com/guides/other-guides/iceberg-tables

It should cover common use cases when using Iceberg tables with external catalogs.

**Reply by leopasta-enable** [reply]
Hi littleK0i, wonder what are your current plans about adding support for Iceberg tables?
It seems snowflake is doubling down on Iceberg as a long term strategy, and now it is generally available.
 
For us, we currently use as an integration mechanism, reading data produced by databricks. So they function very much like an EXTERNAL TABLE for us now.
We're creating the iceberg tables outside snowddl into their own "sandbox" schema, it would be great to have it managed in SnowDDL when the support comes.

A smaller change, that would help us would be to be able to have ICEBERG_TABLES as a permission model object type, this way we could start by having snowDDL managing the schema permissions, instead of doing it in external scripts. 
What do you think?

**Reply by littleK0i** [reply]
Added `ICEBERG_TABLE` object type in `0.34.3`, just to make it available for grants and permission models.

I'll look into possibility to implement actual resolver in the next month or two. The main problem is requirement to set up external storage and all these additional catalog types. Guess we need a separate AWS account and separate testing infrastructure to make it work.

Still not sure if Iceberg brings real value for ~99% of businesses with data volumes up to 1Pb. For very big guys ... maybe, but they probably did their homework and have it in place already, along with custom automation.

**Reply by leopasta-enable** [reply]
Awesome, thanks for that!

In terms of value, I guess the main thing we would be looking for from iceberg is simplifying integration: with the iceberg layer, we can move logic from databricks to snowflake more gradually (and who knows if in future we want to move systems out of snowflake to other stacks :) )

**Reply by littleK0i** [reply]
@leopasta-enable , please note.

SnowDDL now supports unmanaged Iceberg tables starting from version `0.38.0`.

Guide: https://docs.snowddl.com/guides/other-guides/iceberg-tables

Normally it should not affect anything. Iceberg tables start to resolve only if at least one relevant config is present. But I suggest to pay attention in case of updates, just in case.

**Reply by leopasta-enable** [reply]
Awesome Christmas gift, thanks littleK0i and merry Christmas! :)

**Comment by jamesclair** [comment]
Hi @littleK0i, snowflake managed iceberg is starting to get a lot better from interoperability standpoint and we've decided to go that route with a new larger datasource to reduce our ops overhead while keeping raw data sovereignty.  I've thought about contributing an update to support it, but the team is pretty busy recently.  Thought I'd drop a line here in case you are already working on it or want to keep track of the request.

P.S. Thanks for everything you do!

**Reply by littleK0i** [reply]
I see Snowflake now allows "fully managed Iceberg tables" (without integrations, catalogs, etc.), which is great for development and testing. It helps to remove one of the major roadblocks.

The biggest remaining problem is [structured data typing](https://docs.snowflake.com/en/sql-reference/data-types-structured). It does not matter for normal tables, but it is very important for Iceberg.

Unlike normal types, structured types:
- can be nested, with any (?) number of levels;
- can contain object key names with no restrictions (e.g. spaces, special characters, UTF-8);
- can be mixed with each other in any combination;

Parsing, normalisation and validation for structured types seems difficult. I did a few experiments with [Lark](https://docs.snowflake.com/en/sql-reference/data-types-structured). It mostly works, but not sure if it can realistically cover all edge cases and do not cause performance issues.

Another interesting issue with structured data types. Even if we manage to fully cover this in SnowDDL, I have my doubts about other tools. I generally expect a lot of crashing, error or undefined behaviour when other tools encounter deeply nested and complex data types, so I currently suggest to avoid using it for serious production accounts. Of course, actual data structure can still be validated during ETL process.

---

Regarding data sovereignty / data ops. In my view, Iceberg is still a "false promise" for most companies, with very few exceptions. If you decide to move out of Snowflake, unloading large amounts of data (~petabytes) is still quick and cheap. You probably going to spend much more credits on Iceberg overhead relative to "once-in-a-few-years" COPY INTO <location>.

And if you decide to move out, it will most likely involve at least some degree of data transformation. Data engines might be generally compatible with Iceberg, but they still have gimmicks, especially when it comes to large amounts of data and high performance.

**Reply by jamesclair** [reply]
This is interesting and to be honest I'm still very new to iceberg.  There were a few reasons we chose this route v.s. COPY INTO.  We are primarily dealing with a IOT data that uses keys to keep from having high dimensionality.  In otherwords, our data shouldn't be very complex, require nesting or really a high degree of manipulation within the data lake.  But we will need to retrieve large quantities of data from either another layer within snowflake or via a custom ETL hosted on our internal platform.  

The vendor we receive the data from sends it to us in parquet into an s3 bucket.  The data used to be processed on a daily basis, but we are working towards near real-time for a couple of use cases.  We could use mqtt and move some of the cleaning up the pipe, but we don't have the man power to update the processing pipeline to handle the volume at current moment.  The thought was, we could make use of the existing parquet format w/ iceberg, handle archiving directly in S3, and Snowflakes move to make their iceberg api useable by external clients buys us some time to experiment with ETL/ELTs while we shift the processing logic up the pipe.

Considering we still have to update snowflake ingest the data we were definitely on the fence, as we basically pay for the data twice anyways.  But I like that if our code supports iceberg, that it won't need to change anything but the source if we move from using snowflake iceberg in future.  Especially if we can make use of AWS S3 Tables in future which seems like the better route from an IOT and performance stand point.  I think one of our primary goals is staying flexible due to the high rate of change in this area.
